applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_lagrange_dirichlet_condition.cpp

```diff
@@ -0,0 +1,458 @@
+        }
+
+        noalias( rLeftHandSideMatrix ) = ZeroMatrix(matrix_size,matrix_size); //resetting LHS
+    }
+
+    // Resizing as needed the RHS
+    if ( CalculateResidualVectorFlag == true ) //calculation of the matrix is required
+    {
+        if ( rRightHandSideVector.size( ) != matrix_size )
+        {
+            rRightHandSideVector.resize( matrix_size, false );
+        }
+
+        noalias( rRightHandSideVector ) = ZeroVector( matrix_size ); //resetting RHS
+    }
+
+    // Get imposed displacement and normal vector
+    const array_1d<double, 3 > & xg_c = this->GetValue(MPC_COORD);
+    const array_1d<double, 3 > & imposed_displacement = this->GetValue (MPC_IMPOSED_DISPLACEMENT);
+
+    // Prepare variables
+    GeneralVariables Variables;
+    const double augmentation_factor = this->GetValue(SCALAR_LAGRANGE_MULTIPLIER);
+    bool stabilization = this->GetValue(STABILIZATION_LAGRANGE_MULTIPLIER);
+
+    // Calculating shape function
+    Variables.N = this->MPMShapeFunctionPointValues(Variables.N, xg_c);
+    Variables.CurrentDisp = this->CalculateCurrentDisp(Variables.CurrentDisp, rCurrentProcessInfo);
+
+    // Check contact: Check contact penetration: if <0 apply constraint, otherwise no
+    bool apply_constraints = true;
+    if (Is(CONTACT))
+    {
+        // NOTE: the unit_normal_vector is assumed always pointing outside the boundary
+        array_1d<double, 3 > & unit_normal_vector = this->GetValue(MPC_NORMAL);
+        ParticleMechanicsMathUtilities<double>::Normalize(unit_normal_vector);
+        array_1d<double, 3 > field_displacement = ZeroVector(3);
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            if (Variables.N[i] > std::numeric_limits<double>::epsilon() )
+            {
+                for ( unsigned int j = 0; j < dimension; j++)
+                {
+                    field_displacement[j] += Variables.N[i] * Variables.CurrentDisp(i,j);
+                }
+            }
+        }
+
+        const double penetration = MathUtils<double>::Dot((field_displacement - imposed_displacement), unit_normal_vector);
+
+        // If penetrates, apply constraint, otherwise no
+        if (penetration >= 0.0)
+        {
+            apply_constraints = false;
+        }
+    }
+
+    if (apply_constraints)
+    {
+        Matrix lagrange_matrix = ZeroMatrix(matrix_size, matrix_size);
+
+         // Loop over Lagrange Multipliers
+        for (unsigned int i = 0; i < number_of_nodes; i++)
+        {
+            // Loop over shape functions of displacements
+            for (unsigned int j = 0; j < number_of_nodes; j++)
+            {
+                const double NN = Variables.N[j] * Variables.N[i];
+                const unsigned int ibase = i * dimension + dimension * number_of_nodes;
+                const unsigned int jbase = j * dimension;
+
+                // Matrix in following shape:
+                // |aug*H   H^T|
```

Comment:
Which LM approach is this?

Reply:
It is from Belytschko "Nonlinear Finite Element Elements for Continua and Structures" page 350.

Reply:
![imagen](https://user-images.githubusercontent.com/19991680/64591220-97ba1800-d3a9-11e9-93b8-300eeb626035.png)
Indeed, it should be mentioned in the description of the class

Reply:
It should be renamed to ALM instead of LM to avoid confusions
BTW, you wrote:

~~~
// Matrix in following shape:
// |aug*H   H^T|
// |H         0|
~~~

But it is:

~~~
// |lamnda*H + Beta *(G^T G + g H)   G^T|
// |G                                  0|
~~~

Reply:
If the constraints are linear H is zero:
![imagen](https://user-images.githubusercontent.com/19991680/64591721-c1277380-d3aa-11e9-9ca5-19222dcf04b8.png)

So in you case will be, and assuming you are calling G as H:

~~~
// | Beta *(H^T H)   H^T|
// |H                  0|
~~~

Reply:
BTW, there is a typo in the lambda contribution of the residual in 6.3.54 appears as positive and in 63.58 as negative. Be careful

Reply:
Yes I saw the error in the equation 6.3.58 thanks. I am calling G as H but H is N_iN_j as my lagrange multiplier is described via the shape functions and the nodal values. Therefore it results in the form:
| beta*H    H |
|H             0|
as the constraints are linear.

Reply:
See my other comments, update the style and add the corresponding documentation pointing to the theory. In any case there is something wrong if you need to add epsilon values to the diagonal, are you trying to use an iterative solver?, even with that epsilon your condition number will be very ill

Reply:
No I'm not using the iterative solvers. For most cases I use the SuperLU solver. This solver has sometimes difficulties to solve complex problems. If I put the small value on the diagonal than it works. Alternatively I can use the "paradiso" solver for this problem.
For smaller examples like a dynamic cantilever beam with constant point load on the tip it works perfectly and I don't need this stabilization - but it also does not influence my results.
I'll push the changes concerning your comments soon. Thanks for the review :)

Reply:
I see, I think we should study in detail a way to avoid this "stabilization". We can do a short of dual LM as I do in the contact, but in order to do so I need to understand better the MPM formulation in order to adapt this step from the FEM formulation to MPM

Reply:
Ok, perhaps the best way is to delete the "stabilization" from this PR as it is not essential and think about better alternatives like dual LM (which I don't know so far...)